### PR TITLE
convert debug.c to debugprint.d

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/debugprint.d [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+            printf(" b_ret=B%d", b.b_ret ? b.b_ret.Bnumber : 0);
+version (MARS)
+{
+        if (b.Bsrcpos.Sfilename)
+            printf(" %s(%u)", b.Bsrcpos.Sfilename, b.Bsrcpos.Slinnum);
+}
+        printf("\n");
+        if (b.Belem)
+        {
+            if (debugf)
```
Comment (wilzbach) [inline src/dmd/backend/debugprint.d]: First rule of conversion ;-) (it's originally `0`)

File: src/dmd/backend/debugprint.d [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+                        printf(" B%d",list_block(bl).Bnumber);
+                    printf("\n");
+                }
+                break;
+            case BCret:
+            case BCretexp:
+            case BCexit:
+                break;
+            default:
+                printf("bc = %d\n", b.BC);
```
Comment (wilzbach) [inline src/dmd/backend/debugprint.d]: First rule of conversion: this isn't part of the CXX version.

File: src/dmd/backend/debugprint.d [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+        switch (e.Eoper)
+        {   case OPconst:
+                elem_print_const(e);
+                break;
+            case OPrelconst:
+                ferr("#");
+                goto case OPvar;
+
+            case OPvar:
+                printf("%s",e.EV.Vsym.Sident.ptr);
```
Comment (wilzbach) [inline src/dmd/backend/debugprint.d]: This looks a bit dangerous. It doesn't look `char[1]` is null-terminated in the default cause, e.g. https://run.dlang.io/is/VHJOtq

Maybe "%c" or being length-specific with `%.*s`?

File: src/dmd/backend/debugprint.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/backend/debugprint.d]: Eh, must be my copy of debug.c differed slightly. No matter.

File: src/dmd/backend/debugprint.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/backend/debugprint.d]: same as other comment.

File: src/dmd/backend/debugprint.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/debugprint.d]: Sident is always 0 terminated, that's why the original being just Sident worked.

**Review by wilzbach — APPROVED** [review]

**Pull request merged**